core/sr-primitives/src/sr_arithmetic.rs

```diff
@@ -669,16 +669,824 @@ impl CheckedAdd for Fixed64 {
+						true
+					} else {
+						false
+					}
+				};
+
+				test();
+				while (*rhat.borrow() as Double) < B {
+					if !test() { break; }
+				}
+
+				let qhat = qhat.into_inner();
+				// we don't need rhat anymore. just let it go out of scope when it does.
+
+				// step D4
+				let lhs = Self { digits: (j..=j+n).rev().map(|d| self_norm.get(d)).collect() };
+				let rhs = other_norm.clone().mul(Self::from(qhat));
+
+				let maybe_sub = lhs.sub(rhs);
+				let mut negative = false;
+				let sub = match maybe_sub {
+					Ok(t) => t,
+					Err(t) => { negative = true; t }
+				};
+				(j..=j+n).for_each(|d| { self_norm.set(d, sub.get(d - j)); });
+
+				// step D5
+				// PROOF: the `test()` specifically decreases qhat until it is below `B`. conversion
+				// is safe.
+				q.set(j, qhat as Single);
+
+				// step D6: add back if negative happened.
+				if negative {
+					q.set(j, q.get(j) - 1);
+					let u = Self { digits: (j..=j+n).rev().map(|d| self_norm.get(d)).collect() };
+					let r = other_norm.clone().add(u);
+					(j..=j+n).rev().for_each(|d| { self_norm.set(d, r.get(d - j)); })
+				}
+			}
+
+			// if requested, calculate remainder.
+			if rem {
+				// undo the normalization.
+				if normalizer_bits > 0 {
+					let s = SHIFT as u32;
+					let nb = normalizer_bits;
+					for d in 0..n-1 {
+						let v = self_norm.get(d) >> nb
+							| self_norm.get(d + 1).overflowing_shl(s - nb).0;
+						r.set(d, v);
+					}
+					r.set(n - 1, self_norm.get(n - 1) >> normalizer_bits);
+				} else {
+					r = self_norm;
+				}
+			}
+
+			Some((q, r))
+		}
+	}
+
+	#[cfg(feature = "std")]
+	impl rstd::fmt::Debug for Number{
+		fn fmt(&self, f: &mut rstd::fmt::Formatter<'_>) -> rstd::fmt::Result {
+			write!(
+				f,
+				"Number {{ {:?} ({:?})}}",
+				self.digits,
+				u128::try_from(self.clone()).unwrap_or_else(|_| 0),
+			)
+		}
+	}
+
+	impl PartialEq for Number {
+		fn eq(&self, other: &Self) -> bool {
+			// sadly, we have to reallocate here as strip mutably uses self.
+			let mut lhs = self.clone();
+			let mut rhs = other.clone();
+			lhs.lstrip();
+			rhs.lstrip();
+			lhs.digits.eq(&rhs.digits)
+		}
+	}
+
+	impl Eq for Number {}
+
+	impl Ord for Number {
+		fn cmp(&self, other: &Self) -> Ordering {
+			// sadly, we have to reallocate here as strip mutably uses self.
+			let mut lhs = self.clone();
+			let mut rhs = other.clone();
+			// strip both
+			lhs.lstrip();
+			rhs.lstrip();
+			// then resize to max
+			lhs.resize(&mut rhs);
+			// then we can iterate safely
+			let mut r = Ordering::Equal;
+			for (i, p) in lhs.digits.iter().enumerate() {
+				let q = rhs.digits[i];
+				match p.cmp(&q) {
+					Ordering::Less => { r = Ordering::Less; break} ,
+					Ordering::Greater => { r = Ordering::Greater; break} ,
+					_ => continue,
+				}
+			}
+			r
+		}
+	}
+
+	impl PartialOrd for Number {
+		fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+			Some(self.cmp(other))
+		}
+	}
+
+	impl ops::Add for Number {
```

Comment:
Do these traits need to be implemented? I think ideally you want to encourage adds, muls etc. that use references instead of taking ownership. Otherwise looks good.

Reply:
yeah I was being too biased here with my personal preference of taking ownership when an operation always returns. I will make all the rhs's reference.

Reply:
I have to see, maybe we can also not do it here and leave it for some benchmarks to clearly show the difference and then refactor to reference in some future PR.

Reply:
It'd be preferable to choose the API in this PR and leave refactors in future PRs to just optimize implementation without changing API ideally.

Reply:
yes but anyway this is not a big deal. Number implements Clone, changing the API doesn't really break code, user just have to make some more clone and borrow in part of its code that's all.